packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js

```diff
@@ -73,8 +73,14 @@ function isFormElement( element ) {
 export function useFocusFirstElement( clientId ) {
 	const ref = useRef();
 	const initialPosition = useInitialPosition( clientId );
+	const { isBlockSelected } = useSelect( blockEditorStore );
 
 	useEffect( () => {
+		// Check if the block is still selected at the time this effect runs.
```

Comment:
In `trunk`, this code fails to get the focus to the inner block when `templateInsertUpdatesSelection` is `true` because:

- The wrapper is not selected but its `initialPosition` is set, hence this hooks gives it the focus.
- At the point the inner block hook runs it's no longer selected, so it doesn't gain focus. The reason it's no longer selected is because the `useFocusHandler` has already detected that selection & focus are not in sync, hence it [dispatches](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/use-block-props/use-focus-handler.js#L49) the `select_block` action to update selection to the wrapper.

In `trunk`, this code works as expected when `templateInsertUpdatesSelection` is `false` because:

- The wrapper is selected, hence it gets the focus.
- The inner block is not selected (and its initial position is nullish).

Reply:
In `trunk`, or you mean in this branch?

Reply:
I meant in `trunk`, just trying to document the before/after behavior :)

A more precise description of why this fails in `trunk` when `templateInsertUpdatesSelection` is `true` is this:

1. The replace_blocks action is dispatched and the `useFocusFirstElement` runs for the wrapper. At this point, the wrapper is selected and so `initialPosition` is 0. The effect is hooked for React to execute it.
3. Before React executes the effect, the replace_inner_blocks action is dispatched making the inner block selected. At this point the `initialPosition` computation for the wrapper wouldn't be 0. However, the effect for it is already queued and we don't have a way to update that stale info.
4. The effect for the wrapper runs with stale data (`initialPosition` to 0) and focuses the wrapper.
5. The `useFocusHandler` hook detects the wrapper is focused and updates the selection to be in the wrapper again (select_block action).
6. The `useFocusFirstElement` hook runs for the inner block, but the `initialPosition` is undefined. When the effect is executed, it does nothing.

So this fix prevents 4 from focusing the wrapper and 5 from happening. Hence, at 6, the initialPosition for the inner block is 0 and will focus it.

Reply:
I've looked at `useInitialPosition` for other potential issues. It seems the rest are more difficult to get into a race condition like block selection is because they're user actions (happen at lower cycles than redux/react cycles). Though, in theory, it could happen again: what would you think of removing `useInitialPosition` hook entirely and inline its contents within the effect itself for clarity?